### ISSUE TITLE
Turn restrictions

### ISSUE THREAD

**Issue by handcath** [Tristramg/osm4routing2#19]
Hello.
Could you please confirm if osm relations, more specifically turn restrictions, are taken into account? I looked at the code and could not find anything that shows relations are parsed and processed before edges are created. But I may have missed something 😃 .

**Comment by Tristramg** [comment]
Hello,

No there aren’t taken in account. Once you start working with turn restrictions, the model get suddenly way more complicated, and that probably means you are working on some serious stuff, not on a toy project, and osm4routing will be missing tons of little things.

What is you need? All advanced routing engines that use OSM data (Valhalla, OSRM, Graphhopper) support turn restrictions. Maybe you can plug yourself to an intermediate state of their data processing piepeline?

**Comment by handcath** [comment]
Hello and thank you for the clarification.

I have used osrm, and GH. OSRM uses nodes bindings and does not perform well under heavy loads and GH uses Java. So I'm out of luck, so need to go traditional on this, read, build a routing machine myself, which I'll do in Golang.

**Comment by Tristramg** [comment]
I’m somewhat surprised about the performance issues with OSRM. What kind of
routing are you trying to achieve ? I strongly doubt that you could match
osrm in go, the algorithmic part is already very advanced, and osrm has a
very optimized libraries for memory alignment to get the most of cpu caches
(we are talking order of 1000 request/second routing across the globe).

**Comment by handcath** [comment]
I have now tried all of the "CH" open routing machines. OSRM CH is indeed very fast at returning a route (that said annotations are a bit of a performance limiter). I can currently complete thousands of requests/s using the node bindings via distributed docker. Results are sent as they become available to goroutines which add elevation and process a bunch of CPU bound things. It works but I could do without the http overhead.

GH has elevation but the low level lib suffers from a serious lack of documentation, and right or wrong I'm not a Java fan. I have developed on it, but can't replicate OSRMs performance, and GH doesn't expose osm node ids (only edges) unlike OSRM. 

I now have a golang OSM graph generator which adds elevation to node ids split ways by intersection, a router between intersections. I did not build a CH graph, instead I am using transit nodes. Memory consumption goes through the roof but I am not worried about that.
As per your earlier note, relations are tricky to deal with. The key was to route between intersections as opposed to nodes. Transit nodes actually help with that.

I'm no computer scientist, and the subject a bit new to me. I like Go. Pretty barebone but spectacularly within reach for neophytes like me.
